File: web/src/sim/world/ProvinceQuery.ts
import type { WorldData } from '~/sim/data/WorldFile'
import type { Resource } from '~/sim/economy/Resource'
import type { Terrain } from '~/sim/world/Terrain'
import type { WorldState } from '~/sim/world/WorldState'

export interface ProvinceSummary {
  readonly id: number
  readonly ownerName: string
  readonly controllerName: string
  readonly isCity: boolean
  readonly population: number
  readonly terrain: Terrain
  readonly morale: number
  readonly resource: Resource
  readonly isOccupied: boolean
  readonly isContested: boolean
  readonly landNeighbours: number
  readonly seaNeighbours: number
}

/*
 * Semua yang boleh ditanyakan antarmuka tentang satu provinsi, dikumpulkan di
 * satu tempat supaya panel tidak pernah menjangkau larik provinsi sendiri.
 *
 * Nama datang dari luar: berkas biner simulasi tidak menyimpan teks apa pun
 * selain tag negara, dan menambahkannya berarti menaruh teks milik penyaji ke
 * dalam berkas yang sengaja tidak memuat data penyajian.
 */
export class ProvinceQuery {
  readonly #world: WorldState
  readonly #data: WorldData
  readonly #names: string[]

  constructor(world: WorldState, data: WorldData, provinceNames?: readonly string[]) {
    if (data.provinceCount !== world.provinces.count) {
      throw new RangeError(
        `data dunia memuat ${data.provinceCount} provinsi tetapi state memuat ${world.provinces.count}`,
      )
    }

    this.#world = world
    this.#data = data
    this.#names = new Array<string>(world.provinces.count).fill('')

    if (provinceNames) {
      const shared = Math.min(provinceNames.length, this.#names.length)
      for (let i = 0; i < shared; i++) this.#names[i] = provinceNames[i] ?? ''
    }
  }

  get count(): number {
    return this.#world.provinces.count
  }

  nameOf(province: number): string {
    return this.#names[this.#checked(province)]!
  }

  summarise(province: number): ProvinceSummary {
    const p = this.#world.provinces.at(this.#checked(province))

    return {
      id: province,
      ownerName: this.#nationName(p.owner),
      controllerName: this.#nationName(p.controller),
      isCity: p.isCity,
      population: p.population,
      terrain: p.terrain,
      morale: p.morale,
      resource: this.#data.resourceOf(province),
      isOccupied: p.isOccupied,
      isContested: p.isContested,
      landNeighbours: this.#data.land.neighboursOf(province).length,
      seaNeighbours: this.#data.sea.neighboursOf(province).length,
    }
  }

  neighboursOf(province: number): { land: number[]; sea: number[] } {
    this.#checked(province)

    return {
      land: [...this.#data.land.neighboursOf(province)],
      sea: [...this.#data.sea.neighboursOf(province)],
    }
  }

  findCity(name: string): number {
    const wanted = name.trim().toLowerCase()
    if (wanted.length === 0) return -1

    for (let i = 0; i < this.#names.length; i++) {
      if (this.#world.provinces.isCity[i] !== 0 && this.#names[i]!.toLowerCase() === wanted) {
        return i
      }
    }

    return -1
  }

  #nationName(nation: number): string {
    return nation < this.#world.nations.count ? this.#world.nations.name[nation]! : ''
  }

  #checked(province: number): number {
    if (province < 0 || province >= this.#world.provinces.count) {
      throw new RangeError(
        `provinsi ${province} di luar peta 0..${this.#world.provinces.count - 1}`,
      )
    }

    return province
  }
}

File: sim/NationRise.Core/World/ProvinceQuery.cs
using NationRise.Core.Data;
using NationRise.Core.Economy;

namespace NationRise.Core.World;

public readonly record struct ProvinceSummary(
    int Id,
    string OwnerName,
    string ControllerName,
    bool IsCity,
    float Population,
    Terrain Terrain,
    float Morale,
    Resource Resource,
    bool IsOccupied,
    bool IsContested,
    int LandNeighbours,
    int SeaNeighbours);

/*
   Everything the interface may ask about a single province, gathered in one
   place so panels never reach into the province arrays themselves.

   Names arrive from outside: the simulation binary carries no strings at all,
   and adding them to it would put text the renderer owns inside the file that
   deliberately holds no presentation data.
*/
public sealed class ProvinceQuery
{
    private readonly WorldState _world;
    private readonly WorldData _data;
    private readonly string[] _names;

    public ProvinceQuery(WorldState world, WorldData data, IReadOnlyList<string>? provinceNames = null)
    {
        ArgumentNullException.ThrowIfNull(world);
        ArgumentNullException.ThrowIfNull(data);

        if (data.ProvinceCount != world.Provinces.Count)
        {
            throw new ArgumentException(
                $"World data holds {data.ProvinceCount} provinces but the state holds {world.Provinces.Count}.",
                nameof(data));
        }

        _world = world;
        _data = data;
        _names = new string[world.Provinces.Count];
        Array.Fill(_names, string.Empty);

        if (provinceNames is not null)
        {
            int shared = Math.Min(provinceNames.Count, _names.Length);
            for (int i = 0; i < shared; i++)
            {
                _names[i] = provinceNames[i] ?? string.Empty;
            }
        }
    }

    public int Count => _world.Provinces.Count;

    public string NameOf(int province) => _names[Checked(province)];

    public ProvinceSummary Summarise(int province)
    {
        ProvinceRef p = _world.Provinces[Checked(province)];

        return new ProvinceSummary(
            province,
            NationName(p.Owner),
            NationName(p.Controller),
            p.IsCity,
            p.Population,
            p.Terrain,
            p.Morale,
            _data.ResourceOf(province),
            p.IsOccupied,
            p.IsContested,
            _data.Land.NeighboursOf(province).Length,
            _data.Sea.NeighboursOf(province).Length);
    }

    public (IReadOnlyList<int> Land, IReadOnlyList<int> Sea) NeighboursOf(int province)
    {
        Checked(province);
        return (Ids(_data.Land.NeighboursOf(province)), Ids(_data.Sea.NeighboursOf(province)));
    }

    public int FindCity(string name)
    {
        ArgumentNullException.ThrowIfNull(name);

        string wanted = name.Trim();
        if (wanted.Length == 0)
        {
            return -1;
        }

        for (int i = 0; i < _names.Length; i++)
        {
            if (_world.Provinces.IsCity[i]
                && string.Equals(_names[i], wanted, StringComparison.OrdinalIgnoreCase))
            {
                return i;
            }
        }

        return -1;
    }

    private static int[] Ids(ReadOnlySpan<ushort> neighbours)
    {
        var ids = new int[neighbours.Length];
        for (int i = 0; i < neighbours.Length; i++)
        {
            ids[i] = neighbours[i];
        }

        return ids;
    }

    private string NationName(ushort nation) =>
        nation < _world.Nations.Count ? _world.Nations.Name[nation] : string.Empty;

    private int Checked(int province)
    {
        ArgumentOutOfRangeException.ThrowIfNegative(province);
        ArgumentOutOfRangeException.ThrowIfGreaterThanOrEqual(province, _world.Provinces.Count);
        return province;
    }
}
